### DISCUSSION TITLE
Bug with `consult-fd`: `error: Found argument '--and' which wasn't expected, or isn't valid in this context`

### DISCUSSION THREAD

**Discussion by hab25** [minad/consult#1049] [Q&A] [answered]
Environment: emacs 29.3, consult 1.7, fd 10.1.0

Reproduce by evaluating in `emacs -Q`:
```
(progn
  (require 'consult)
  (minibuffer-with-setup-hook
      (apply-partially #'insert "foo")
    (consult-fd)))
```

And seeing that `error: Found argument '--and' which wasn't expected, or isn't valid in this context` appears in the minibuffer and in the `*Messages*` buffer.

**Comment by minad** [comment]
Hi! This problem is caused if your `fd` version is too old.

**Reply by hab25** [reply]
Hi, could there be another cause here? `10.1.0` is the latest tagged release; the `--and` flag was introduced before that, see https://github.com/sharkdp/fd/blob/4e9672250b3866a8bc9d4bd0129349ecd81cf9df/CHANGELOG.md?plain=1#L105C3-L106C21.

**Reply by minad** [reply]
I don't know. For me `consult-fd` works properly with multiple inputs. I also have version 10.something. Maybe you've adjusted the `consult-fd-args`?

**Reply by hab25** [reply]
> Maybe you've adjusted the consult-fd-args?

No, I've reproduced it `emacs -Q` and evaluating only the above code. 

---

Another test we can compare: 

Again in `emacs -Q`, I did `(debug-on-entry #'consult--async-log)`, then called `#'consult-fd`, inputted the string "bar", and the debugger landed on the following:

`consult--async-log("consult--async-process started %S\n" ("fdfind" "--full-path" "--color=never" "--and" "bar" "."))`

Then, on the command line of my terminal (outside of Emacs), I invoked `fdfind --full-path --color=never --and bar .` and got the same error:
`error: Found argument '--and' which wasn't expected, or isn't valid in this context`

**Reply by minad** [reply]
> Then, on the command line of my terminal (outside of Emacs), I invoked fdfind --full-path --color=never --and bar . and got the same error:
error: Found argument '--and' which wasn't expected, or isn't valid in this context

Yes, this means that your version of fd is not supported. This command line works on my setup, version 9.0.0. It could be that fd 10.0 got more strict regarding the usage of the --and argument. But I haven't found anything in the change log. See also this problem: https://github.com/minad/consult/commit/e201af1404097c1cb6ef36bd250e81269e0cfb71

If you have time you could test various fd versions and check how the command line should be organized, such that we can pass multiple patterns and such that patterns are not confused with options or file names.

**Reply by maxecharel** [reply, marked as answer]
Hi! For what it's worth, I encountered the same error message as the one encountered by @hab25. In my case, the problem was due to what @minad suggested, i.e. that the version of `fd` was too old (8.3.1, package `fd-find`). I installed a newer pre-packaged (`.deb`) version (9.0.0), and it solved the issue. 

@hab25, I may be completely wrong, but is it possible that the implementation of `fd` called by `consult-fd` is actually not the one that corresponds to the output of `fd --version`? On some Debian-based distros, the package available on the official repos is `fd-find` and the related command is `fdfind`, not `fd`. Could it be that two versions --- e.g. an old `fdfind` from the official repos and a `fd` that you built --- coexist on your current system and that `consult-fd` actually calls `fdfind` while you would like it to call `fd`? 

I am not sure that's even possible and I don't know whether you use a Debian-based distro, but let's hope that sharing these thoughts can help.

**Reply by hab25** [reply]
@maxecharel that is indeed the cause; impressive guess; thanks!

The solution I went with, because my distro (pop os 22.04) has the `fdfind` binary as a dependency of the desktop environment and I shouldn't uninstall it, was to "substitute" (i.e. via `setopt`) `https://github.com/minad/consult/blob/3d0fc6a8c6a74b21ca854b8632b3f58e0e513b85/consult.el#L291C5-L291C58` with `"fd"` (which I have installed through the `nix` package manager).

**Reply by maxecharel** [reply]
hi @hab25, glad it helped!
